### ISSUE TITLE
retrieve the chain mail of thread

### ISSUE THREAD

**Issue by goseivietnam** [jstedfast/MailKit#1254]
Good morning team, i stuck with get chaining mail in folder inbox, for example in my gmail :

Customer send me a mail : hello Nam
i answer: hello Customer
then customer answer again : hello Nam again
as i expected, i will have one thread and three mail in that thread
mail 1 [hello Nam] 
mail 2 [hello Customer], isreply = true
mail 3 [hello Nam again], isreply = true

but follow answer of this question in stack overflow
https://stackoverflow.com/questions/61122028/using-mailkit-to-retrieve-email-chains-returns-only-the-first-email-of-the-conv
i see the thread start with mail 1 [hello Nam]  and mail 2 [hello Customer] only, what i wrong to select mail 3 ?
or do you have any example for me to collect all mail send or received in thread ?
this function same gmail when you click on thread of inbox folder

**Comment by jstedfast** [comment]
- What threading algorithm are you using?
- Are you using MessageThreader with a list of IMessageSummary objects? Or are you using ImapFolder.Thread()?
- Are you threading a list of messages that includes all 3 messages? Remember that MessageThreader needs all 3 messages in order to thread them. It can't pull message #3 out of thin air.
- If you are using MessageThreader, what MessageSummaryItems flags are you passing to Fetch()? Are you sure you have all of the necessary info?
- If you are threading using the `References` algorithm, does the third message have an `In-Reply-To` header or a `References` header? Do the values in those headers reference the Message-Id of message 2?

**Comment by IwishIcanFLighT** [comment]
Hi!

I know this is an old issue but I figured my situation is similar to OP's so I might as well ask here...

I'm trying to list an email thread/conversation kinda like what Gmail does:

```
[Mail 1] Hello there
[Mail 2] Re: Hello there
[Mail 3] Re: Hello there
```

From what I found in my tests with MailKit and googling around, mails 1 and 3 will end up in the Inbox folder and mail 2 will be in the Sent folder.
This is where I'm having difficulties to understand and use `MessageThreader.Thread()` because it needs a list of summaries to work with, which means that I have to `Fetch()` both Inbox and Sent folders if I want the full email list.

Furthermore, what if one of the emails was moved to another folder, say mail 1 is in moved in `MyCustomArchiveFolder` ? The only option I see is to `Fetch()` every folders on the account which seems very costly in IMAP calls.

Any help is welcome, thanks!

**Comment by jstedfast** [comment]
@IwishIcanFLighT are you using GMail as an IMAP server in your case? Or is this a question about "how to do it for a general IMAP account"?

The second is MUCH MUCH MUCH harder.

**Comment by IwishIcanFLighT** [comment]
Well, both I suppose. 😅  In the long run, I want my project to support both Gmail and any IMAP account.

**Comment by jstedfast** [comment]
The first thing to understand is that GMail uses a proprietary threading algorithm.

Next, if you Fetch() the GMailThreadId, you can group messages by the MessageSummary.GMailThreadId to get GMail's message grouping. So that's easy. Oh, and you can use the `client.GetFolder (SpecialFolder.All)` which gives you a reference to GMail's AllMail folder which is has every message in the account (Inbox, Sent, etc).

For a typical IMAP server, you'll have to collect the summary info for every folder and feed it into MessageThreader.Thread() as you've already discovered. You'd probably want to use the `ThreadingAlgorithm.References` algorithm even though that isn't identical to GMail's algorithm, but it's the closest algorithm that MailKit implements.

In order to use the References algorithm, you want to make sure to use `MessageSummaryItems.UniqueId | MessageSummaryItems.Envelope | MessageSummaryItems.References` at a minimum in your Fetch() request.

**Comment by ekalchev** [comment]
We built our threading algorithm using information from here.

https://www.jwz.org/doc/threading.html

**Comment by jstedfast** [comment]
@ekalchev FWIW, that's the `ThreadingAlgorithm.References` algorithm. The algorithm that JWZ describes in that document is what was later adopted as the THREAD=REFERENCES IMAP threading algorithm.

**Comment by IwishIcanFLighT** [comment]
@jstedfast Thanks for your detailed reply. It confirms what I thought... It will end up rather costly in IMAP calls unless I cache as much as possible all the summaries of the account. Maybe with a local sqlite DB.

I also noted in my findings that `IMessageSummary.References` seems to be a list of `UniqueID` corresponding to each emails in the thread. Since I'll have to get all the account summaries anyway, I might as well create my own threading algorithm using the `References`.

So to summarize:
1. Get all the account summaries, cache them to avoid constant huge IMAP calls
2. When attempting to rebuild an email's thread, fetch the emails...
 a) Prioritize searching the Inbox and Sent folder, the most likely folders where the emails I'm looking for are
 b) If nothing, Then lookup in other folders (user archived the emails in folders)
 c) If still nothing, try to renew the cache and attempt a) & b) again
 d) If _still_ nothing, I only have the `UniqueID` in `References` to work with (user probably deleted the email)
3. Display my final email list of the thread in the UI.

This is a bit overkill, but it provides a very intuitive list that I can display in my UI. I'm not surprised that Gmail's threading algorithm is proprietary. Their threaded display is very good and appealing for the end-user compared to most email clients.

**Comment by jstedfast** [comment]
@IwishIcanFLighT `IMessageSummary.References` isn't a list of `UniqueIds`, it's a list of Message-ID's.

> It will end up rather costly in IMAP calls unless I cache as much as possible all the summaries of the account. Maybe with a local sqlite DB.

Correct.

> Since I'll have to get all the account summaries anyway, I might as well create my own threading algorithm using the References.

You could, but I'm not sure I understand why you are planning to do so. MessageThreader can already thread based on the References.

**Comment by IwishIcanFLighT** [comment]
> @IwishIcanFLighT IMessageSummary.References isn't a list of UniqueIds, it's a list of Message-ID's.

Yes my bad. I often get confused between the two.

> You could, but I'm not sure I understand why you are planning to do so. MessageThreader can already thread based on the References.

My plan is to have my UI display a full email conversation like Gmail does. Apparently it's called "Conversation View" as I just found out, sorry if it confused you earlier. It basically looks like this:

![](https://it.stonybrook.edu/sites/default/files/kb/5613/images/conversation-view-collapsed-annotated-for-first-and-newest-message.jpg)

I tested `MessageThread.Thread()` and it works great, but ultimately it builds a nested object that won't work with my UI that will have a ListView. What I actually need is a sorted by date List of `IMessageSummary` and I'm good to go.
